### PR TITLE
Fix distutils.LooseVersion DeprecationWarning

### LINKED ISSUE #84712: Fix deprecation warning in `torch.utils.tensorboard` import
### 🚀 The feature, motivation and pitch

Importing the tensorboard writer yields a deprecation warning from `distutils` in python 3.10, i.e.,
`DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.`

This can be easily fixed as mentioned in the deprecation warning. One needs to replace the current lines 2 - 9 (https://github.com/pytorch/pytorch/blob/master/torch/utils/tensorboard/__init__.py#L2)
```
from distutils.version import LooseVersion

if not hasattr(tensorboard, "__version__") or LooseVersion(
    tensorboard.__version__
) < LooseVersion("1.15"):
    raise ImportError("TensorBoard logging requires TensorBoard version 1.15 or above")

del LooseVersion
```

with 

```
import packaging

if not hasattr(tensorboard, "__version__") or packaging.version.parse(
        tensorboard.__version__
) < packaging.version.Version("1.15"):
    raise ImportError("TensorBoard logging requires TensorBoard version 1.15 or above")

del packaging
```

I am happy to submit the PR but I do not know how to test against all supported python versions and changing all requirements correctly in all different build targets. 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Jasha10**
Fixes #84712

### CODE REVIEW

**Comment by Jasha10** [comment]
@milutter, I tried opening a PR based on the diff from your post on issue #84712.
I'm seeing some test failures: `module 'packaging' has no attribute 'version'`

```text
Traceback (most recent call last):
  File "test_monitor.py", line 104, in setUp
    from torch.utils.tensorboard import SummaryWriter
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/tensorboard/__init__.py", line 4, in <module>
    if not hasattr(tensorboard, "__version__") or packaging.version.parse(
AttributeError: module 'packaging' has no attribute 'version'
```

(here's an example CI job failure: https://github.com/pytorch/pytorch/actions/runs/3397514945/jobs/5649878533#step:10:3456)

**Comment by MaKaNu** [comment]
the module `version` needs to be imported explicitly. The Devs from "neutron" had a [similar issue](https://bugs.launchpad.net/neutron/+bug/1904854).

Simply

```python

from packaging import version

...
if not hasattr(tensorboard, "__version__") or version.parse(
        tensorboard.__version__
) < version.Version("1.15"):
...

```

could fix this.

File: torch/utils/tensorboard/__init__.py [hunk 1]
```diff
@@ -1,12 +1,12 @@
 import tensorboard
-from distutils.version import LooseVersion
+import packaging
```
Comment (MaKaNu) [inline torch/utils/tensorboard/__init__.py:2]: change to `from packaging import version`

File: torch/utils/tensorboard/__init__.py [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
 
-if not hasattr(tensorboard, "__version__") or LooseVersion(
-    tensorboard.__version__
-) < LooseVersion("1.15"):
+if not hasattr(tensorboard, "__version__") or packaging.version.parse(
```
Comment (MaKaNu) [inline torch/utils/tensorboard/__init__.py:4]: change `packaging.version.parse` to `version.parse`

File: torch/utils/tensorboard/__init__.py [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
+        tensorboard.__version__
+) < packaging.version.Version("1.15"):
```
Comment (MaKaNu) [inline torch/utils/tensorboard/__init__.py:6]: change `packaging.version.Version("1.15")`to `version.Version("1.15")`

File: torch/utils/tensorboard/__init__.py [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
     raise ImportError("TensorBoard logging requires TensorBoard version 1.15 or above")
 
-del LooseVersion
+del packaging
```
Comment (MaKaNu) [inline torch/utils/tensorboard/__init__.py:9]: change `packaging` to `version`

**Review by MaKaNu — CHANGES_REQUESTED** [review]

**Comment by Jasha10** [comment]
Thanks @MaKaNu! I've updated the PR based on your comments.

**Comment by milutter** [comment]
Thanks everybody for the effort! The explicit import of `version`, is a bit weird but seems to work. Is the current problem that `packaging` is not listed as a dependency in https://github.com/pytorch/pytorch/blob/master/requirements.txt? 

Can we try adding it to the requirements?

**Comment by Jasha10** [comment]
Thanks for the tip, @milutter! I've added `packaging` to `requirements.txt`.

**Comment by MaKaNu** [comment]
@milutter  I wondered what may caused the necessity to explicit import `version`. I think the `AttributeError` gives a good hint:

**module** 'packaging' has not attribute 'version'

but which module is meant? Packaging is a package which next to other modules the `__init__.py` and `version.py` includes. So if we import `packaging` only we don't import the package instead we import the package as module, which results in the `__init__.py`. Since the module `__init__` only has the choosen attributes of the module `__about__.py` and the attribute `__all__`, it does not know anything about the module `version.py`.

If we instead call `from packaging import version` , we reference now on the package directly.

**Comment by milutter** [comment]
Thanks @MaKaNu for the explanation. 

I looked a bit deeper into resolving the problem. I still believe this is a requirements problem such that MYPY does not find the types stubs. Adding it to `requirements.txt` is false because this `requirements.txt` is only for development. I personally could not find where the dependencies need to be added. 

To get it past the `pylint` one could try the really hacky version: 
```
if not hasattr(tensorboard, "__version__") or tuple(int(x) for x in tensorboard.__version__.split("."))[:2] < (1, 15):
```

**Comment by Jasha10** [comment]
> To get it past the pylint one could try the really hacky version:

Hopefully we can get eyes on this PR from someone who knows how the lints work.

**Comment by MaKaNu** [comment]
I ran a small test env and used the package as torch used it. Worked fine and ran mypy manuall over it -> No Errors. I am not completly sure about it, but I will find it again, but this might be a wrong configured mypy setup: I don't have the link at the moment, but will search it, but I have read about stubs and things for builtin packages. If stubs are absolutley necessary aswell for builtin, we need to add them. Stay tuned for further informations tomorrow.

**Comment by MaKaNu** [comment]
I looked into it again and found the conversations, but I do not understand completely: [https://github.com/python/mypy/issues/638](https://github.com/python/mypy/issues/638). On the other hand I found another issue [https://github.com/python/mypy/issues/4542](https://github.com/python/mypy/issues/4542), which is pointing to the issue, which also mention [PEP 561](https://peps.python.org/pep-0561/) as a solution. I did not had the time to dig deeper into this, but this might help to find the issue.

**Comment by milutter** [comment]
I looked at the links from @MaKaNu and they seem to be related. The different search path between python and MyPy seem like a big mess. Even when reading the complete posts, I do not fully understand the problem or see a possible solution. From other blogpost, I have seen two hacks that might mitigate the problem (https://stackoverflow.com/questions/68695851/mypy-cannot-find-implementation-or-library-stub-for-module):

1) Import the complete function, i.e., `from packaging.version import parse`
2) Tag line with `# type: ignore` to disable MyPY for this line.

@Jasha10 Would you be able to try these different solutions?

**Comment by Jasha10** [comment]
Looks like the `# type: ignore` comment worked.

**Review by milutter — APPROVED** [review]
Thanks for trying and fixing it. I am happy to get this finally resolved. 

Do you know the process of actually merging this PR? Does somebody within PyTorch need to approve & merge the PR. How does one get their attention for this?

**Comment by MaKaNu** [comment]
I don't know but hope, that now as every checks have passed an authorized user get informed about it. At the end I still wonder why mypy even has a problem with the builtin lib. I think something is not correctly configured for mypy, since a non configured mypy install does not react on this.

**Review by MaKaNu — APPROVED** [review]

File: torch/utils/tensorboard/__init__.py [hunk 5]
```diff
@@ -1,12 +1,12 @@
 import tensorboard
-from distutils.version import LooseVersion
+from packaging import version  # type: ignore[import]
```
Comment (MaKaNu) [inline torch/utils/tensorboard/__init__.py:2]: still think the ignore solves the mypy issue but should not.

**Review by MaKaNu — APPROVED** [review]

File: torch/utils/tensorboard/__init__.py (hunk 5 above)
Comment (MaKaNu) [reply torch/utils/tensorboard/__init__.py:2]: I don't know why I have looked into, but there are actually **3** `mypy.ini` configs. I understand why we use `mypy.ini` and `mypy-strict.ini`. But the uncommented `mypy-nofollow.ini`? From the commit message I got a hint. 

Which of them is used in this pipeline? I don't know but it should be the normal `mypy.ini` or `mypy-nofollow.ini` but not the strict one.

**Comment by Jasha10** [comment]
> At the end I still wonder why mypy even has a problem with the builtin lib.

FYI @MaKaNu, `packaging` is not built-in:

```python
$ python -q
>>> import packaging
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'packaging'
```

**Comment by MaKaNu** [comment]
> FYI @MaKaNu, packaging is not built-in

this is weird and you are right. Somehow on my wsl installation packaging is installed without environment.

**Comment by milutter** [comment]
@Jasha10 I just saw that you are part of the PyTorch organization, would you be able to get some movement into this PR, such that it is reviewed and accepted?

**Comment by Jasha10** [comment]
@milutter I'm **honored** to be a part of the pytorch org, but TBH I don't know _why_ I'm part of the pytorch org. (I was added to it randomly 🤷🏼)

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
eh, I just realized I can't merge this, as `packaging` is not an inbuilt (I just saw the comment from @MaKaNu above).

Can we use something that doesn't add a dependency?

**Review by soumith — CHANGES_REQUESTED** [review]
shouldn't be using packaging

**Comment by soumith** [comment]
we could add `packaging` as a dependency too, but it isn't added in as a dependency right now in this PR

**Comment by Jasha10** [comment]
We can parse the version number by hand:
```python
>>> import tensorboard
>>> tuple(map(int, tensorboard.__version__.split(".")))
(2, 11, 0)
```

The advantage of `packaging.version.parse` is that it's robust against different styles of version strings, e.g. it could handle `"v1.2.3.dev0"`.

**Comment by milutter** [comment]
thanks @soumith for taking a look at this and trying to merge the PR! 

I would prefer to add `packaging` as a dependency because this package is recommend by the deprecated python system package `distutils` and avoiding this package would result in very ugly code (and it would not be very generic code or handle edge-cases which are hard to enumerate in this case). 

What seems weird to me that the tests are passing when packaging is not a dependency?
In an initial draft the CI failed on the packaging import (see https://github.com/pytorch/pytorch/pull/88524#issuecomment-1304389816). More precisely, `test-reports/python-unittest/test_public_bindings` failed with `AttributeError: module 'packaging' has no attribute 'version'` (I copied the full output below). Once the import got fixed CI passed, which points in the direction that there is a rudimentary CI test that should fail when the packaging package is not installed. So somehow `packaging` could already be a dependency of Pytorch?! 

Nevertheless, where are the python dependencies listed in pytorch? I looked into the classical [`requirements.txt`](https://github.com/pytorch/pytorch/blob/master/requirements.txt), but this states that it only contains the `dev` dependencies.


Older CI Failure:
```
 Test results will be stored in test-reports/python-unittest/test_public_bindings
  test_correct_module_names (__main__.TestPublicBindings)
An API is considered public, if  its  `__module__` starts with `torch.` ... ERROR (3.636s)
  test_correct_module_names (__main__.TestPublicBindings)
An API is considered public, if  its  `__module__` starts with `torch.` ...     test_correct_module_names errored - num_retries_left: 3
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_public_bindings.py", line 387, in test_correct_module_names
    for _, modname, ispkg in pkgutil.walk_packages(path=torch.__path__, prefix=torch.__name__ + '.'):
  File "/opt/conda/lib/python3.10/pkgutil.py", line 107, in walk_packages
    yield from walk_packages(path, info.name+'.', onerror)
  File "/opt/conda/lib/python3.10/pkgutil.py", line 92, in walk_packages
    __import__(info.name)
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/tensorboard/__init__.py", line 4, in <module>
    if not hasattr(tensorboard, "__version__") or packaging.version.parse(
AttributeError: module 'packaging' has no attribute 'version'
```

**Comment by soumith** [comment]
you should add `packaging` as a dependency here and we're good: https://github.com/pytorch/pytorch/blob/master/setup.py#L980-L983

The reason CI passes is because the CI is pre-populated with a bunch of packages (for a variety of reasons)

**Comment by Jasha10** [comment]
Done!

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks for the fix @Jasha10 !

**Comment by milutter** [comment]
@weiwangmeta: This PR got reverted by your PR [#91548](https://github.com/pytorch/pytorch/pull/91548). All the test failures link to internal meta resources, which make it hard to debug. Could you copy the test failure error message to investigate the failure?

**Comment by soumith** [comment]
yea the failures are around having to adjust internal build files. one of us will have to take it on. Can you reopen the PR

**Comment by soumith** [comment]
actually, looks like @Jasha10 is at Meta. Jascha, you can open this PR as a diff in FBCode and add `packaging` as a dependency to the caffe2/TARGETS file. then once you put up the diff, click "export to Github" in the UI, then all will be well to land it.

**Comment by Jasha10** [comment]
@soumith I am no longer at Meta. Sorry I couldn't get to this while I was still there.

**Pull request closed without merging**